### DISCUSSION TITLE
Creating distro packages for Photons (and friends)

### DISCUSSION THREAD

**Discussion by Djelibeybi** [delfick/photons#36] [Ideas]
Motivated by #33, this discussion is to share tips and tricks for creating RPM or DEB packages that install Photons. 

I'm currently noodling on creating a set of RPMs, based on the official [Fedora Packaging Guidelines](https://docs.fedoraproject.org/en-US/packaging-guidelines/). I'll need to package [`delfick_project`](https://github.com/delfick/delfick_project) to provide the dependency, but I think the rest are already available in Fedora.

My intention is to create three RPMs: `python3-lifx-photons-core`, `python3-lifx-photons-arranger` and `python3-lifx-photons-interactor`. The `lifx` command-line utility is provided by `python3-lifx-photons-core` while the two app RPMs will depend on the core one and will include `systemd` service unit files which meet the Fedora guidelines.

**Comment by joshua-pe** [comment]
Sounds cool. Hopefully i can switch  the image over to this when it's done. Currently using standard interactor install (no docker) with a systemd service running for photons-image, I couldn't figure out how to do what you are doing now. 

One annoying thing i just discovered about the github mobile app is that doesn't support discussions, Oh well

**Reply by delfick** [reply]
You can probably reply via email like PRs and issues (I deliberately not logged into github on my phone to try and make me not interact with it when I'm not at a computer, but I can still reply to things using emails)

**Comment by delfick** [comment]
probably also a good time to figure out switching from travis ci github actions and generate rpms with that once you have the ability to make those.

**Reply by Djelibeybi** [reply]
I find GitHub Actions easier to understand than Travis but YMMV. 🤷‍♀️ GitHub Actions certainly have more powerful flexibility with less effort, given how many Actions have already been written. And I'm _very_ lazy, so I appreciate that.

**Reply by delfick** [reply]
yeah, I've wanted to migrate for a while now, but time. I think I have time today :)

**Reply by delfick** [reply]
Ok, new core/arranger/interactor released. All using github actions! and will now generate tags and github releases.

Will be super simple to emit rpms/debs in that once we have code that can generate them

**Comment by joshua-pe** [comment]
Hi Guys!

Thought I would give you an update on photons imager. I have just finished creating a webpage for controlling the animations so people can use it on all platforms. Hoping to get the image up ~the 8th (that's ~when the rpi is supposed to arrive)

Here's a screenshot of the webpage, done all in python! Still need to iron out some bugs and add some more of the animations.

![image](https://user-images.githubusercontent.com/63589370/103456683-470bc280-4d44-11eb-8445-1e2a674f3d3b.png)

**Comment by delfick** [comment]
cool. got a github repo somewhere?

**Reply by joshua-pe** [reply]
Yup, it’s not uploaded yet, just the documentation with no doubt a couple of typos. Also need to add documentation for the web controller https://github.com/Final-Hawk/Photons-Image

**Reply by delfick** [reply]
oh yeah. So you've gone down the route of manually burning an SD card to generate the image? I'm all for you creating your own rPi distribution and I encourage that, but I just wanna gently push you in a slightly different direction.

I would suggest giving it it's own name ( that doesn't say photons) as it's a distribution that happens to have photons on it combined with your bonus stuff. Maybe I'll pull in the ability to photons to generate such a thing, but that's unlikely to happen any time soon, and you're including an operating system here :) 

I would split your code up:
* The codebase of the stuff you're adding
* The codebase/process of what you use to create the image
* The image itself

Also, I would discourage using wifi with an rPi and make people plug it into Ethernet!

And, finally, I'm not a fan of random .exe files with no source code. It's gonna be less work for you and more trustworthy for users to supply an .img and suggest using [etcher](https://www.balena.io/etcher/)

**Reply by joshua-pe** [reply]
Thanks for the suggestions. I haven’t actually made my own distribution, it’s just raspberry Pi os lite with the host name changed. What should I do with that? Just keep it as “raspberry Pi”? 

With the Ethernet thing I would definitely prefer that and I will include that in the instructions, but the raspberry zero doesn’t have a Ethernet jack without a micro USB to Ethernet adapter. 

With the exe I will change that and just get people to download etcher. Good point on the security thing. 

The way I am creating the image is just installing raspberry Pi os on a non expanded 8gb sd card, then installing photons with a systemd service to start it. Then I am simply using the command line to clone and export it as an image. Nothing too hard. Thanks for the feedback!

**Reply by delfick** [reply]
I didn't know the zero had no ethernet port. How strange.

Names are hard, I'm sure you can find a fun name for this project though :)

I just want to make it explicit as to what it is and how you've made it, which is a hand made raspberry pi image that happens to use photons to give you tile animations.

**Reply by joshua-pe** [reply]
Alright thanks, I will make a wiki page sometime today with how I made it, than on the readme I will just say, something like “how does it work and how did I make it” and link to the wiki. Something like that just to be more open. 

I’ll see what name I can come up with :)

**Reply by joshua-pe** [reply]
Just finished the wiki on what it runs on. What do you think? https://github.com/Final-Hawk/Photons-Image/wiki/What-does-the-image-contain%3F That github actions sound good if it makes rpm's and debs easier.

**Reply by delfick** [reply]
nice :)

You should add a "What is this?" section to the readme that explains it's a raspberry Pi distribution that sets up photons, hand made by you and point to that page in the wiki for more info.

And with a fancy name once you rename it, sounds good :)

**Reply by joshua-pe** [reply]
Good idea on the what's this, don't know how i overlooked that. 

I have come up with a generic name (Light Magic) but it will do until i think of something else.

**Reply by delfick** [reply]
excellent.

(I already said this, but the email reply sent as a new comment rather than in the thread, maybe I need to press reply all next time)

**Comment by joshua-pe** [comment]
So I have been messing around and I cannot seem to get photons interactor to run via a systemd service. It keeps on reporting that it failed to start the interactor. For the initial release I will just use docker I suppose, although it is unnecessary. Look out for a reddit post this afternoon if everything goes to plan

**Reply by delfick** [reply]
what does `sudo journalctl -fu interactor` say? (assuming your service is called interactor)

**Reply by joshua-pe** [reply]
It returns this: Although if I run the ExecStart command manually it works. All of the directories are there as well. 

Jan 05 08:26:30 raspberrypi systemd[767]: photons.service: Failed to execute command: No such file or directory

Jan 05 08:26:30 raspberrypi systemd[767]: photons.service: Failed at step EXEC spawning /home/pi/photons/bin/lifx: No such file or 
directory

Jan 05 08:26:30 raspberrypi systemd[1]: photons.service: Main process exited, code=exited, status=203/EXEC


Jan 05 08:26:30 raspberrypi systemd[1]: photons.service: Failed with result 'exit-code'.


Jan 05 08:26:31 raspberrypi systemd[1]: photons.service: Service RestartSec=100ms expired, scheduling restart.


Jan 05 08:26:31 raspberrypi systemd[1]: photons.service: Scheduled restart job, restart counter is at 5.


Jan 05 08:26:31 raspberrypi systemd[1]: Stopped The Photons interactor.


Jan 05 08:26:31 raspberrypi systemd[1]: photons.service: Start request repeated too quickly.


Jan 05 08:26:31 raspberrypi systemd[1]: photons.service: Failed with result 'exit-code'.


Jan 05 08:26:31 raspberrypi systemd[1]: Failed to start The Photons interactor.

**Reply by delfick** [reply]
` Failed at step EXEC spawning /home/pi/photons/bin/lifx: No such file or directory`

**Reply by joshua-pe** [reply]
Yes that's the issue, i just found out that something is wrong and its not finding any files whatsoever. Now i think of it it could be permissions

**Reply by delfick** [reply]
what does `ls -Rl /home/pi/photons` say? (put it in a gist, it'll probably be a lot)

**Reply by joshua-pe** [reply]
Ok so still the same error. Heres the output of "ls -Rl /home/pi/.photons-core"
https://pastebin.com/a9S5AzVw

**Reply by delfick** [reply]
so instead of `/home/pi/photons/bin/lifx` you want `/home/pi/.photons-core/bin/lifx`

**Reply by joshua-pe** [reply]
Yea, thats what it is. I reinstalled so it changed back sorry. This is the current systemd service. For the moment i will just use docker though. Its not much of a hassle


[Unit]
Description=The Photons interactor
After=network.target
Before=graphical.target

[Service]
ExecStart='/home/.photons-core/bin/lifx lan:interactor'
WorkingDirectory=/home/pi/.photons-core/
Environment=LIFX_CONFIG=/home/.photons-core/lifx.yml
KillMode=process
Restart=on-failure

[Install]
WantedBy=multi-user.target

**Reply by delfick** [reply]
and the journalctl error still says it can't find it?

I would suggest putting lifx.yml outside the virtualenv too.

So perhaps `/home/pi/.config/photons/lifx.yml`

So make that and then run this from the command line and see if running it manually works

```
cd /home/pi/.config/photons
LIFX_CONFIG=/home/pi/.confg/photons/lifx.yml /home/.photons-core/bin/lifx lan:interactor
```

Also, your problem is that you have quotes for ExecStart  so it's trying to find a file called `lifx lan:interactor` rather than a file called `lifx`

You want 

``` 
[Unit]
Description=The Photons interactor
After=network.target
Before=graphical.target

[Service]
ExecStart=/home/.photons-core/bin/lifx lan:interactor
WorkingDirectory=/home/pi/.conf/photons
Environment=LIFX_CONFIG=/home/pi/.conf/photons/lifx.yml
KillMode=process
Restart=on-failure

[Install]
WantedBy=multi-user.target
```

**Reply by joshua-pe** [reply]
I need to stop and go to bed i think. I am so sorry for wasting your time on an error I made that I kept overlooking. I'll see if you can spot it. (its in execstart)

One i was using:
[Unit]
Description=The Photons interactor
After=network.target
Before=graphical.target

[Service]
ExecStart=/home/.photons-core/bin/lifx lan:interactor
WorkingDirectory=/home/pi/configdir/
Environment=LIFX_CONFIG=/home/pi/configdir/lifx.yml
KillMode=process
Restart=on-failure

[Install]
WantedBy=multi-user.target

One i needed to use :
[Unit]
Description=The Photons interactor
After=network.target
Before=graphical.target

[Service]
ExecStart=/home/pi/.photons-core/bin/lifx lan:interactor
WorkingDirectory=/home/pi/configdir/
Environment=LIFX_CONFIG=/home/pi/configdir/lifx.yml
KillMode=process
Restart=on-failure

[Install]
WantedBy=multi-user.target

**Reply by delfick** [reply]
hhehehe, oops :p 

And yes, me to, bedtime!

**Comment by joshua-pe** [comment]


**Reply by delfick** [reply]
I do not unfortunately.

**Comment by joshua-pe** [comment]


**Reply by delfick** [reply]
cool. you should rename the repo and change references to the new name in the readme :)

**Reply by joshua-pe** [reply]
Done! Still trying to keep most references here though.

**Reply by delfick** [reply]
:)

**Comment by delfick** [comment]
@Djelibeybi you can do a pip freeze to generate a "lock file" as such.

```
python3 -m venv photons
./photons/bin/pip install lifx-photons-core lifx-photons-interactor
./photons/bin/pip freeze
```

Which should be fine for creating a package

**Reply by delfick** [reply]
but yes, gemfile and package-lock are nice and python packaging is a nightmare. (also everytime I look at upcoming python packaging alternatives like pipenv or poetry I don't like them, so I stick with setup.py, pip and venvstarter for now)

**Reply by Djelibeybi** [reply]
> @Djelibeybi you can do a pip freeze to generate a "lock file" as such.

Yep, I've just discovered `pipdeptree` and `freeze` via that. Work continues afoot!

**Reply by delfick** [reply]
woot

**Reply by delfick** [reply]
oooh, pipdeptree looks cool

**Comment by Djelibeybi** [comment]
RPMs are now available on COPR for Fedora Rawhide, Fedora 33 and EL8-derivatives for both `x86_64` and `aarch64`: https://copr.fedorainfracloud.org/coprs/djelibeybi/photons/

Currently being built from a temporary repo at https://github.com/djelibeybi/photons-rpms because it was easier to deal with just the packaging bit first. I'll look at merging at least the Photons bits into this repo sometime in the future, once Fedora has all the required Python modules, etc.

**Comment by delfick** [comment]
cool. do you have scripts that generate these? so I can add to github
actions

**Comment by Djelibeybi** [comment]
No, COPR does that itself. I need to work out the right/most efficient way to integrate the actual source repo and COPR, which I’ll look at next week.

**Reply by delfick** [reply]
you don't need extra things?

**Reply by Djelibeybi** [reply]
> you don't need extra things?

I don't _think_ so. But getting this far took way more effort than I expected, so I'm taking a wee break before coming back to it.

**Comment by delfick** [comment]
fair enough. so what exactly was involved?

**Reply by Djelibeybi** [reply]
> fair enough. so what exactly was involved?
> […](#)

I wanted to make sure that I was adhering to the Fedora Packaging Guidelines, which meant learning more about systemd and how it now manages system users, for example. Also spent a bunch of time on making sure it worked with any version of Python, from 3.6 (the default on EL8) to 3.9 on Fedora Rawhide.

I also had to disable a whole bunch of automatic dependency generation because some of the  versions you used are not packaged upstream. In those cases, I either repackaged a newer version or just overrode Photons' explicit dependency with a >= if upstream were newer. 

https://docs.fedoraproject.org/en-US/packaging-guidelines/

**Comment by delfick** [comment]
oh yeah. so there are additional files involved.
